ibis/expr/types/numeric.py

```diff
@@ -480,13 +481,15 @@ def sqrt(self) -> NumericValue:
         """
         return ops.Sqrt(self).to_expr()
 
+    @util.deprecated(instead="use nullif(0)", as_of="7.0", removed_in="8.0")
     def nullifzero(self) -> NumericValue:
         """Return `NULL` if an expression is zero."""
-        return ops.NullIfZero(self).to_expr()
+        return self.nullif(ops.Literal(0, dtype=self.type()).to_expr())
```

Comment:
I can see why we need to be careful with casting the 0 with fillna(0) vs fillna(0.0) so you end up with the right type.in the result, but here we are just testing for equality so zeros of any type will still be equal, right? Eg ibis.literal(0).nullif(0.0) will still work? As expected?

If so, I would say remove this extra cast.

Reply:
It needs to stay to propagate the type information in case a backend ever needs to insert an actual cast.

And also the use of ops is to avoid having to import Ibis.

Reply:
There won't be a cast inserted unless the backend does it specifically. The literal construction is merely propagating type information.

Reply:
Sorry if I'm missing something here. Or maybe I wasn't clear enough, I didn't just mean, "remove the cast", I meant the entire implementation should just be simply `return self.nullif(0)`. If we can't, that implies the migration path that we suggest to users is flawed. Maybe can you give an example of why `return self.nullif(0)` wouldn't work? Thanks!

Reply:
I don't think I follow why this implementation detail would change the behavior of the result and change any equivalence between what was previously happening. It may not be necessary, but if it's implemented this way, I don't have to worry about whether all databases' notion of equality match python's

Reply:
Ah, it wouldn't change in relation to previous behavior. I'm saying that if we are suggesting users to switch from `zeroifnull()` to `fillna(0)`, then we want those two to be equivalent. If we switch the implementation here to just be a wrapper around `fillna(0)`, and we don't change the .zeroifnull() tests at all (except for adding the pytest.warns...), then we are actively testing for that equivalence. If we make the implementation here slightly different, then the behavior between `.fillna(0)` and `zeroifnull()` could be slightly different, our tests won't catch it, users will switch over to `.fillna(0)`, and their results will change. We WANT our tests to fail and to know if databases' have different notions of equality, because then that means that the migration path from `zeroifnull()` to `.fillna(0)` is trickier than we thought, and we need to add in something extra.

Reply:
In fact, per that argument, the implementation for `nullifzero()` should also be simply `nullif(0)`, to verify that clean migration for that method.

Reply:
Ok, I see what you're saying! Makes sense. It's possible that `fillna` of `ifnull` may need to be modified to make the recommendation work :)